concent_api/core/models.py

```diff
@@ -73,3 +77,236 @@ class MessageAuth(Model):
+    report_computed_task        = models.OneToOneField(StoredMessage, blank = True, null = True, related_name = 'subtasks_for_report_computed_task')
+    force_get_task_result       = models.OneToOneField(StoredMessage, blank = True, null = True, related_name = 'subtasks_for_force_get_task_result')
+    ack_report_computed_task    = models.OneToOneField(StoredMessage, blank = True, null = True, related_name = 'subtasks_for_ack_report_computed_task')
+    reject_report_computed_task = models.OneToOneField(StoredMessage, blank = True, null = True, related_name = 'subtasks_for_reject_report_computed_task')
+    subtask_results_accepted    = models.OneToOneField(StoredMessage, blank = True, null = True, related_name = 'subtasks_for_subtask_results_accepted')
+    subtask_results_rejected    = models.OneToOneField(StoredMessage, blank = True, null = True, related_name = 'subtasks_for_subtask_results_rejected')
+
+    def __init__(self, *args, **kwargs):
+        super(Subtask, self).__init__(*args, **kwargs)
+        if self.pk is not None:
+            self._current_state = self.state
+
+    def clean(self):
+        super(Subtask, self).clean()
+
+        # next_deadline can be None only for passive states
+        if self.pk is not None and self.next_deadline is None and self.state not in self.SUBTASK_PASSIVE_STATES:
+            raise ValidationError({
+                'next_deadline': 'next_deadline is None for active state.'
+            })
+
+        # State transition can happen only by defined rule
+        # but not when we create new object
+        # and not when state is not being changed
+        if (
+            self.pk is not None and
+            self._current_state != self.state and
+            self._current_state not in self.POSSIBLE_STATE_TRANSITIONS[self.state]
+        ):
+            raise ValidationError({
+                'state': 'Subtask cannot change its state from {} to {}.'.format(
+                    self._current_state, self.state
+                )
+            })
+        else:
+            self._current_state = self.state
+
+
+class PaymentCommittedMessage(models.Model):
+    """
+    Stores messages with payments for requestor to be returned from the 'receive-out-of-band' endpoint
+    """
+
+    class RecipientType(ChoiceEnum):
+        Provider    = 'provider'
+        Requestor   = 'requestor'
+
+    payment_ts              = models.DateTimeField()
+    task_owner_key          = Base64Field(max_length = 64)
+    provider_eth_account    = Base64Field(max_length = 64)
+    amount_paid             = models.FloatField()
+    recipient_type          = models.CharField(max_length = 32, choices = RecipientType.choices())
+    amount_pending          = models.FloatField()
+
+
+class PendingResponse(models.Model):
+    """
+    Stores information about messages to be returned from the `receive` or `receive-out-of-band` endpoint.
+    """
+
+    class ResponseType(ChoiceEnum):
```

Comment:
This list does not match the one from blueprint. Here it contains messages that are never supposed to be returned from `receive/` but wrapped in other messages.

Reply:
`SubtaskResultsAccepted` can be changed to `ForceSubtaskResultsResponse` like in blueprint. I will fix it.
`ForceGetTaskResult` - I see there is a separate discussion about this one so I am leaving it for now.
`ForceSubtaskResultsSettled` - I think this one is missing in blueprint.
`AckReportComputedTask` and `RejectReportComputedTask` - As far as I remember there is a corner case for that, that we can't set `ForceReportComputedTaskResponse` there and base on subtask`s `ack_report_computed_task` and `reject_report_computed_task`, because in one case we want to send back `AckReportComputedTask` when `RejectReportComputedTask` was received from client.

Reply:
> `ForceGetTaskResult` - I see there is a separate discussion about this one so I am leaving it for now.

I do not see this discussion in comments here. What is the issue with `ForceGetTaskResult`? Do you mean that the upload message is missing from golem-messages? Has it been reported? 

> `AckReportComputedTask` and `RejectReportComputedTask` - As far as I remember there is a corner case for that, that we can't set `ForceReportComputedTaskResponse` there and base on subtask's `ack_report_computed_task` and `reject_report_computed_task`, because in one case we want to send back `AckReportComputedTask` when `RejectReportComputedTask` was received from client.

You don't have to look **only** at the presence of ack or reject in the subtask table. `receive/` should look at the rejection reason, deadline and report timestamp and decide whether to return a reject or an ack.

Reply:
> ForceSubtaskResultsSettled - I think this one is missing in blueprint.

There's no such message in Łukasz's spec or in golem-messages. There's only `SubtaskResultsSettled` and it's been listed in my blueprint from the beginning.